### PR TITLE
build: Add Craft for streamlined releases

### PR DESCRIPTION
**Pull request by BYK**
First step of APP-12.

The next step will be creating a GitHub Action to trigger the builds from there.

~~This now depends on getsentry/craft#379 for proper per-project changelogs.~~

### CODE REVIEW

File: .craft.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            'yarn',
+            '--no-git-tag-version',
+            'version',
+            '--new-version',
+            process.env.CRAFT_NEW_VERSION,
+          ].join(' '),
+          { cwd: package }
+        ).toString()
+      )
+      .join('\n');
```
Comment (lizraeli) [inline .craft.yml:22]: Do we need to map packages (e.g. as opposed to `forEach`) if they don't get returned?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -19,15 +19,13 @@ Shared tsconfig for use across all projects.
 First, install all the dependencies (using [Volta](https://volta.sh/) is highly recommended):
 
 ```shell
-yarn install
+yarn install --frozen-lockfile
```
Comment (lizraeli) [inline CONTRIBUTING.md:22]: Haha I was wondering about the difference between `yarn install` and `yarn install --frozen-lockfile`, and found a github thread you commented on from 2017
https://github.com/yarnpkg/yarn/issues/4147#issuecomment-321774630

File: README.md [hunk 3]
```diff
@@ -14,6 +14,16 @@ React component to load formsort flows.
...
+## Publishing
+
+This repo is published as a single unit with all packages at the same version. It uses [Craft](https://github.com/getsentry/craft) for publishing.
+
+You need to set the following environment variables to be able to work with Craft:
+
+1. `GITHUB_TOKEN`: [Create a personal access token](https://github.com/settings/tokens/new?scopes=repo&description=Craft) from GitHub with `repo` permissions and set the token value to this env variable
+2. `NPM_TOKEN`: Create an automation token from NPM by visiting https://www.npmjs.com/settings/<npm_user>/tokens/new and select **Automation** as the token type.
+
+Then all you need to do is `yarn release <version>`. We'll take it from here :)
```
Comment (lizraeli) [inline README.md:26]: 🎉

File: packages/custom-question-api/examples/react/src/index.tsx [hunk 4]
```diff
@@ -35,11 +39,11 @@ const MyCustomComponent = () => {
...
-        value={name}
+        value={name.toString()}
         onChange={(e) => setName(e.target.value)}
       />
     </div>
   );
 };
 
-ReactDom.render(<MyCustomComponent />, document.getElementById("root"));
+ReactDom.render(<MyCustomComponent />, document.getElementById('root'));
```
Comment (lizraeli) [inline packages/custom-question-api/examples/react/src/index.tsx:49]: I would put these changes in a separate PR

File: packages/custom-question-api/src/WindowMessageEmitter.ts [hunk 5]
```diff
@@ -17,7 +17,10 @@ class WindowMessageEventsEmitter extends EventEmitter {
 
   onWindowMessage = (e: MessageEvent) => {
     const { type, payload, requestId } = e.data;
-    if (!type || !EVENTS_TO_EMIT.has(type)) {
+    if (
+      typeof type !== 'string' ||
+      !EVENTS_TO_EMIT.has(type as CustomQuestionMessage)
+    ) {
```
Comment (lizraeli) [inline packages/custom-question-api/src/WindowMessageEmitter.ts:23]: I think this should go in a separate PR

File: packages/react-embed/src/index.tsx [hunk 6]
```diff
@@ -32,27 +33,25 @@ export interface IReactEmbedEventMap {
...
 ): void => {
-  Object.entries(events).forEach(([reactEventName, listener]) => {
+  for (const [reactEventName, listener] of Object.entries(events)) {
     const embedEventName =
       eventMapping[reactEventName as keyof IReactEmbedEventMap];
-    embed.addEventListener(embedEventName, listener);
-  });
+    embed.addEventListener<typeof listener>(embedEventName, listener);
+  }
 };
```
Comment (lizraeli) [inline packages/react-embed/src/index.tsx:56]: Also these changes should be in a separate PR

**Review by lizraeli — COMMENTED** [review]
The changes for release and setup process look good! 
I think the changes to code should go in a separate PR(s)

**Comment by BYK** [comment]
>The changes for release and setup process look good!
>I think the changes to code should go in a separate PR(s)

Thanks a lot @lizraeli! These changes are due to the setup process and unified linting/type checking. Our code had lots of broken types and builds. I can try to separate them out but then the "release system" PR would only consist of the readme and `yarn release` script addition. Not sure if that would be really useful.

WDYT?

File: packages/react-embed/README.md [hunk 7]
```diff
@@ -2,11 +2,11 @@
 
 Embed [Formsort](https://formsort.com) flows within react components
 
-### Installation
+## Installation
 
-Add @formsort/react-embed to your project by executing `npm install @formsort/react-embed` or `yarn add @formsort/react-embed`.
+Add @formsort/react-embed to your project by executing `yarn add @formsort/react-embed` or `npm install @formsort/react-embed`.
```
Comment (alikirec) [inline packages/react-embed/README.md:7]: yarn promotion detected 😄

Comment (alikirec) [inline packages/react-embed/README.md:7]: yarn promotion detected 😄

File: packages/web-embed-api/src/.npmignore [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Ignore everything in the root
```
Comment (alikirec) [inline packages/web-embed-api/src/.npmignore:1]: I think root `.npmignore` already takes care of this.

**Comment by lizraeli** [comment]
> > The changes for release and setup process look good!
> > I think the changes to code should go in a separate PR(s)
> 
> Thanks a lot @lizraeli! These changes are due to the setup process and unified linting/type checking. Our code had lots of broken types and builds. I can try to separate them out but then the "release system" PR would only consist of the readme and `yarn release` script addition. Not sure if that would be really useful.
> 
> WDYT?

Ah got it. I'll approve. Next time could you put this kind of reasoning in the PR description - that would save some of the back-and-forth. Thanks!

File: packages/custom-question-api/examples/react/src/index.tsx [hunk 9]
```diff
@@ -1,28 +1,32 @@
...
-          setName("");
+          setName('');
         }
         return;
       }
       setAnswerValue(name);
     };
 
-    updateName();
+    void updateName();
```
[the remark this answers is not in the document]
Comment (lizraeli) [reply packages/custom-question-api/examples/react/src/index.tsx:29]: Yeah also not sure why `void` is needed here

[the remark this answers is not in the document]
Comment (BYK) [reply packages/custom-question-api/examples/react/src/index.tsx:29]: It was enforced by a lint rule to ensure we were not returning anything from this function. Agree that it is not great but cannot remove it unless I disable linting here (which I don't really want to)

File: packages/web-embed-api/src/.npmignore (hunk 8 above)
Comment (BYK) [reply packages/web-embed-api/src/.npmignore:1]: Yeah, this should have been removed. Will remove.

File: README.md [hunk 10]
```diff
@@ -14,6 +14,16 @@ React component to load formsort flows.
 
 Embed flows within other webpages, allowing communication between the embed and the parent page.
 
+## Publishing
+
+This repo is published as a single unit with all packages at the same version. It uses [Craft](https://github.com/getsentry/craft) for publishing.
```
[the remark this answers is not in the document]
Comment (BYK) [reply README.md:19]: You won't be publishing a new version of `web-embed-api` by its own anymore, you'll be publishing all the packages at once. Yes it may yield to bumps with no changes but it helps you to not hunt the matching versions for our packages. It also makes many things regarding publishing simpler.

If this becomes a problem we can revisit it in the future so I'd prefer to move forward with this unless you have strong objections.

File: .craft.yml (hunk 1 above)
Comment (BYK) [reply .craft.yml:22]: This is to get the output as `node -p` prints the result of the last expression and in `trace` logging level, Craft emits these for debugging.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lizraeli) [reply README.md:19]: Is there no way for it to auto-detect which packages were changed and only publish those? That's what lerna used to do, so it feels a bit like a step back

[the remark this answers is not in the document]
Comment (BYK) [reply README.md:19]: Craft doesn't deal with that. We can certainly do this, even use lerna for this part. The issue is, we'll be creating GitHub releases for each release and since this is a monorepo, it is easier for end users to track changes in each version change.

[the remark this answers is not in the document]
Comment (lizraeli) [reply README.md:19]:
```suggestion
The issue is, we'll be creating GitHub releases for each release and since this is a monorepo, it is easier for end users to track changes in each version change.
```
Do you mean it's easier if we _do_ only publish the changed packages, or if we publish and increase the version of each package every time? 

To me the idea of all packages pinned to the same version doesn't seem ideal. If I understand correctly, this would mean that a major version change to e.g. `web-embed-api` would also require the same version change to `custom-question-api`, even thought the two packages are independent of each other, and even if the latter had no changes to it.

[the remark this answers is not in the document]
Comment (lizraeli) [reply README.md:19]: To me the idea of all packages being pinned to the same version doesn't seem ideal. If I understand correctly, this would mean that a major version change to, e.g. `custom-question-api`, would require to the make the same version change to `web-embed-api`, even though the two packages are completely independent of each other, and even if the latter had no changes to it.

[the remark this answers is not in the document]
Comment (BYK) [reply README.md:19]: >If I understand correctly, this would mean that a major version change to, e.g. custom-question-api, would require to the make the same version change to web-embed-api, even though the two packages are completely independent of each other, and even if the latter had no changes to it.

That is true yes. My counter argument is that the scenario you provide seems unlikely for the current state of the repo: if there's a breaking change in one of these packages, it is very likely that either other packages will have a breaking change too (like an api change) or one of the packages they both depend on such as `@formsort/constants` has a breaking change.

[the remark this answers is not in the document]
Comment (lizraeli) [reply README.md:19]: These are the current versions of the packages:
- constants: 1.6.0
- custom-question-api: 0.1.0
- eslint-config: 1.3.7
- tsconfig: 1.0.2
- web-embed-api: 2.2.0
- react-embed: 2.1.0

With this change, would we need to update all of these to be the same version (e.g. 2.2.1)?

[the remark this answers is not in the document]
Comment (BYK) [reply README.md:19]: >With this change, would we need to update all of these to be the same version (e.g. 2.2.1)?

Yes, exactly. I think that's a good thing but since you raised the point and are concerned, I think we should find a [Stalemate Resolution Associate](http://www.portal2sounds.com/28). Pinging @fzembow and @atikenny for them to weigh in.

[the remark this answers is not in the document]
Comment (lizraeli) [reply README.md:19]: Yup, I still have my concerns, but if @atikenny and @fzembow feel okay about this I'll be on board too

[the remark this answers is not in the document]
Comment (fzembow) [reply README.md:19]: I definitely do **not** want to be pushing major versions of sibling packages for unrelated reasons.... changelogs with major changes that have no notes associated with them? That's not a good developer experience.

Custom question API and web embed api have almost nothing to do with one another, besides perhaps their use of the same constants. The main reason these are in a monorepo in the first place is because of these more subtle relationships between the packages.

If the consequence of this is -- well, they should be in different repositories, not in a monorepo, then I'm fine with that, although that makes development more difficult.

[the remark this answers is not in the document]
Comment (BYK) [reply README.md:19]: We can certainly separate the deploys while still in a monorepo with separate changelogs. The only issue is we won't be able to use GitHub Releases properly as they are scoped to the repo.

You can see some examples from here: https://github.com/getsentry/sentry-ruby/releases?page=5

Another downside is one having to manually set which repo they are releasing from the UI.

All that being said, since I am 2 to one, I'll yield and set this up for per-package deployment for now.

**Review by alikirec — APPROVED** [review]

File: README.md [hunk 11]
```diff
@@ -14,6 +14,16 @@ React component to load formsort flows.
...
+## Publishing
+
+This repo is published using [Craft](https://github.com/getsentry/craft).
+
+You need to set the following environment variables to be able to work with Craft:
+
+1. `GITHUB_TOKEN`: [Create a personal access token](https://github.com/settings/tokens/new?scopes=repo&description=Craft) from GitHub with `repo` permissions and set the token value to this env variable
+2. `NPM_TOKEN`: Create an automation token from NPM by visiting https://www.npmjs.com/settings/<npm_user>/tokens/new and select **Automation** as the token type.
+
+Then all you need to do is `yarn workspace @formsort/<package> run release <version>`. We'll take it from here :)
```
Comment (lizraeli) [inline README.md:26]: should this be run from the root of the project?

**Review by lizraeli — APPROVED** [review]

File: README.md (hunk 11 above)
Comment (BYK) [reply README.md:26]: Yes (or you can run `yarn release <version>` from the package sub-directory. I'll add a note for this.

Comment (BYK) [reply README.md:26]:
```suggestion
Then all you need to do is `yarn workspace @formsort/<package> run release <version>` from the repo root or `yarn release <version>` from the package directory you want to release. We'll take it from here :)
```
